### PR TITLE
refactor: extract server app

### PR DESCRIPTION
**Pull request by GregoryAlbouy**

### CODE REVIEW

File: runner/internal/config/config.go [hunk 1]
```diff
@@ -124,36 +124,33 @@ func (cfg Global) String() string {
 // Only fields specified in options are replaced. Accepted options are limited
 // to existing Fields, other values are silently ignored.
 func (cfg Global) Override(c Global) Global {
-	if len(c.fieldsSet) == 0 {
-		return cfg
-	}
-	for field := range c.fieldsSet {
+	for field := range cfg.fieldsSet {
```
Comment (c1-ra) [inline runner/internal/config/config.go:130]: I preferred a little more the `prev.Override(next)` order, but I guess either one is ok as long as it is clear in the comment describing the function.

File: runner/internal/config/config.go [hunk 2]
```diff
@@ -120,6 +121,13 @@ func (cfg Global) String() string {
...
 }
 
+// Equal returns true if cfg and c are equal configurations.
+func (cfg Global) Equal(c Global) bool {
+	cfg.fieldsSet = nil
+	c.fieldsSet = nil
+	return reflect.DeepEqual(cfg, c)
+}
+
 // Override returns a new Config based on cfg with overridden values from c.
```
Comment (c1-ra) [inline runner/internal/config/config.go:126]: I think it's the opposite now?
`// Override returns a new Config based on c with overridden values from cfg`

File: httpclient/handle.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+func handle(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+
+	b, err := io.ReadAll(r.Body)
+	if err != nil {
+		internalError(w, err)
+		return
+	}
+
+	cfg, err := configparse.JSON(b)
```
Comment (c1-ra) [inline httpclient/handle.go:24]: Isn't it possible to have yaml here too?

**Review by c1-ra — COMMENTED** [review]
Nice work!
I just have one comment but everything else seems great.

File: httpclient/port.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+	if err != nil {
+		return "", err
+	}
+
+	listener, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return "", err
+	}
+
+	return strconv.Itoa(listener.Addr().(*net.TCPAddr).Port), nil
```
Comment (moreirathomas) [inline httpclient/port.go:19]: With the new data flow, the listener is not reused by `http.Server(listener)` but instead a new one is created via `http.Server.ListenAndServe(port)`.

We should then close this one here when we're done with it.

```suggestion
	listener, err := net.ListenTCP("tcp", addr)
	if err != nil {
		return "", err
	}
	defer listener.Close()

	return strconv.Itoa(listener.Addr().(*net.TCPAddr).Port), nil
```

**Review by moreirathomas — CHANGES_REQUESTED** [review]
`httpclient` seems very logical. Would you go down the road of embedding `engine` into `desktop` and calling the client via `server` there ?

I think we may rebase main here (I suspect you just didn't have the time yet :)).

I also asked for a change on `FindFreePort`.

File: runner/internal/config/config.go (hunk 2 above)
Comment (moreirathomas) [reply runner/internal/config/config.go:126]: It is outdated doc, this has been done on the pr dealing with extracting the cli.

This pr should be rebase on main, as only the very last commit is relevant :)

**Comment by GregoryAlbouy** [comment]
> Would you go down the road of embedding `engine` into `desktop` and calling the client via `server` there ?

I stopped here wondering what would be the next step. I opened the PR anyway for visibility & discuss the changes but I'm not even sure any of these has to be merged here.

How about moving the source code for server to `desktop` in `src-server` (alongside `src` and `src-tauri`) for instance?

**Comment by moreirathomas** [comment]
> How about moving the source code for server (i.e. cmd + httpclient + modules stuff) to desktop in src-server (alongside src and src-tauri) for instance?

This may be the way to go, we are vendoring anyway.


For now I suggest that in the scope of this pr `httpclient` and `server` stay inside `engine`.

And then we can work on the changes on `desktop`. What do you think?

**Review by moreirathomas — APPROVED** [review]

**Comment by GregoryAlbouy** [comment]
Let's merge the PR as an improvement step then, but we definitely need to get `server` out of here so this `engine` lib remains truly agnostic of its consumers (e.g. gh action dispatch to desktop really shouldn't be here)

**Comment by moreirathomas** [comment]
Totally agree given the direction we're going in. Small step is fine :)

**Pull request merged**